utils/cmake/dreamcast.toolchain.cmake

```diff
@@ -0,0 +1,106 @@
+set(CMAKE_SYSTEM_VERSION 1)
+set(PLATFORM_DREAMCAST TRUE)
+
+##### Configure Cross-Compiler #####
+set(CMAKE_CROSSCOMPILING TRUE)
+set(CMAKE_C_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-gcc)
+set(CMAKE_CXX_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-g++)
+
+set(CMAKE_EXECUTABLE_SUFFIX ".elf")
+
+set(CMAKE_FIND_LIBRARY_SUFFIXES ".a")
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+
+##### Add Platform-Specific #defines #####
+ADD_DEFINITIONS(
+    -D__DREAMCAST__
+    -D_arch_dreamcast
+)
+
+if($ENV{KOS_SUBARCH} MATCHES naomi)
+    ADD_DEFINITONS(
+        -D__NAOMI__
+        -D_arch_sub_naomi
+    )
+else()
+    ADD_DEFINITIONS(-D_arch_sub_pristine)
+endif()
+
+##### Configure Build Flags #####
+add_compile_options(-ml -m4-single-only -ffunction-sections -fdata-sections)
+
+set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DFRAME_POINTERS -mfsrra -mfsca -fno-omit-frame-pointer")
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DFRAME_POINTERS -mfsrra -mfsca -fno-omit-frame-pointer")
+
+set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -mfsrra -mfsca -fomit-frame-pointer")
+set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -mfsrra -mfsca -fomit-frame-pointer")
+
+set(CMAKE_C_FLAGS_MINSIZEREL "${CMAKE_C_FLAGS_MINSIZEREL} -mfsrra -mfsca -fomit-frame-pointer")
+set(CMAKE_CXX_FLAGS_MINSIZEREL "${CMAKE_CXX_FLAGS_MINSIZEREL} -mfsrra -mfsca -fomit-frame-pointer")
+
+set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} -DFRAME_POINTERS -mfsrra -mfsca -fno-omit-frame-pointer")
+set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -DFRAME_POINTERS -mfsrra -mfsca -fno-omit-frame-pointer")
+
+set(CMAKE_ASM_FLAGS "")
+set(CMAKE_ASM_FLAGS_RELEASE "")
+
+##### Configure Include Directories #####
+set(CMAKE_SYSTEM_INCLUDE_PATH "${CMAKE_SYSTEM_INCLUDE_PATH} $ENV{KOS_BASE}/include $ENV{KOS_BASE}/kernel/arch/dreamcast/include $ENV{KOS_BASE}/addons/include $ENV{KOS_BASE}/../kos-ports/include")
+
+INCLUDE_DIRECTORIES(
+    $ENV{KOS_BASE}/include
+    $ENV{KOS_BASE}/kernel/arch/dreamcast/include
+    $ENV{KOS_BASE}/addons/include
+    $ENV{KOS_BASE}/../kos-ports/include
+)
+
+##### Configure Libraries #####
+set(CMAKE_SYSTEM_LIBRARY_PATH "${CMAKE_SYSTEM_LIBRARY_PATH} $ENV{KOS_BASE}/addons/lib/dreamcast $ENV{KOS_PORTS}/lib")
+
+if($ENV{KOS_SUBARCH} MATCHES naomi)
+    set(CMAKE_EXE_LINKER_FLAGS " -ml -m4-single-only -Wl,-Ttext=0x8c020000 -Wl,--gc-sections -T$ENV{KOS_BASE}/utils/ldscripts/shlelf-naomi.xc -nodefaultlibs" CACHE INTERNAL "" FORCE)
+else()
+    set(CMAKE_EXE_LINKER_FLAGS " -ml -m4-single-only -Wl,-Ttext=0x8c010000 -Wl,--gc-sections -T$ENV{KOS_BASE}/utils/ldscripts/shlelf.xc -nodefaultlibs" CACHE INTERNAL "" FORCE)
+endif()
+
+LINK_DIRECTORIES(
+    $ENV{KOS_BASE}/addons/lib/dreamcast
+    $ENV{KOS_PORTS}/lib
+)
+
+add_link_options(-L$ENV{KOS_BASE}/lib/dreamcast -L$ENV{KOS_BASE}/addons/lib/dreamcast)
+LINK_LIBRARIES(-Wl,--start-group -lstdc++ -lkallisti -lc -lgcc -Wl,--end-group m)
+LINK_LIBRARIES(c gcc kallisti stdc++)
```

Comment:
I don't think we actually need this last line. I can't find anything that fails to link when removing it and just seems to add the flags multiple times. The start and end group should be enough as far as I can tell.